### PR TITLE
Deprecate mbedtls_cipher_setup_psa()

### PR DESCRIPTION
**Pull request by mprse**
## Description
Deprecate mbedtls_cipher_setup_psa()

Resolves: https://github.com/Mbed-TLS/mbedtls/issues/5261

## Status
**READY**

### CODE REVIEW

**Comment by mpg** [comment]
The CI is unhappy. I guess you need to add a few `#if !defined(MBEDTLS_DEPRECATED_REMOVED)` / `depends_on:!MBEDTLS_DEPRECATED_REMOVED` in tests.

**Comment by mprse** [comment]
Yes I saw that I need to go back to this one, but I started working on something else and I would like to finish at least some part of the new task first.

**Comment by mpg** [comment]
Sure, no worries! I was just checking for new things I should be reviewing, and then writing down the status. Nothing urgent here.

**Comment by mprse** [comment]
Fixed dependencies in tests. Let's see if CI is happy now.

**Comment by mpg** [comment]
CI not quite happy yet, but only one component failing, so this can be tested locally: `tests/scripts/all.sh test_full_no_deprecated_deprecated_warning`.

**Comment by gilles-peskine-arm** [comment]
> In this case we have mbedtls_cipher_setup_psa() available, so I assume that test should be executed, but because MBEDTLS_DEPRECATED_WARNING got deprecation warnings forced to errors.

Gate that part of the tests with `#if !defined(MBEDTLS_TEST_DEPRECATED)`. See [comment in `component_test_full_deprecated_warning`](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-2.28.0/tests/scripts/all.sh#L1481).

**Comment by mpg** [comment]
> Gate that part of the tests with `#if !defined(MBEDTLS_TEST_DEPRECATED)`.

Looking at the code I think it should be the opposite: calls to deprecated functions in tests should be guarded by `#if defined(MBEDTLS_TEST_DEPRECATED)` (no `!`). Or am I missing something?

File: tests/suites/test_suite_cipher.aes.data [hunk 1]
```diff
@@ -2015,243 +2015,243 @@ depends_on:MBEDTLS_AES_C:MBEDTLS_CIPHER_MODE_CBC
 test_vec_crypt:MBEDTLS_CIPHER_AES_256_CBC:MBEDTLS_DECRYPT:"0000000000000000000000000000000000000000000000000000000000000000":"00000000000000000000000000000000":"5c9d844ed46f9885085e5d6a4f94c7d7":"014730f80ac625fe84f026c60bfd547d":0:0
 
 AES-128-CBC crypt Encrypt NIST KAT #1 PSA
-depends_on:MBEDTLS_USE_PSA_CRYPTO:MBEDTLS_AES_C:MBEDTLS_CIPHER_MODE_CBC
+depends_on:MBEDTLS_USE_PSA_CRYPTO:!MBEDTLS_DEPRECATED_REMOVED:MBEDTLS_AES_C:MBEDTLS_CIPHER_MODE_CBC
```
Comment (mpg) [inline tests/suites/test_suite_cipher.aes.data:2018]: I think the extra dependency should be `MBEDTLS_TEST_DEPRECATED` (see comment in the test function).

File: tests/suites/test_suite_cipher.function [hunk 2]
```diff
@@ -1100,9 +1104,9 @@ void test_vec_crypt( int cipher_id, int operation, data_t *key,
 
 exit:
     mbedtls_cipher_free( &ctx );
-#if defined(MBEDTLS_USE_PSA_CRYPTO)
+#if defined(MBEDTLS_USE_PSA_CRYPTO) || defined(MBEDTLS_DEPRECATED_REMOVED)
```
Comment (mpg) [inline tests/suites/test_suite_cipher.function:1107]: Shouldn't we have the same guard here as for the call to `psa_crypt_init()`, that is `USE_PSA_CRYPTO && !DEPRECATED_REMOVED && TEST_DEPRECATED`? (Or just `USE_PSA_CRYPTO && TEST_DEPRECATED`?)

File: tests/suites/test_suite_cipher.function [hunk 3]
```diff
@@ -119,7 +119,8 @@ static int cipher_reset_key( mbedtls_cipher_context_t *ctx, int cipher_id,
     mbedtls_cipher_free( ctx );
     mbedtls_cipher_init( ctx );
 
-#if !defined(MBEDTLS_USE_PSA_CRYPTO)
+#if !defined(MBEDTLS_USE_PSA_CRYPTO) || defined(MBEDTLS_DEPRECATED_REMOVED) || \
+    !defined(MBEDTLS_TEST_DEPRECATED)
```
Comment (mpg) [inline tests/suites/test_suite_cipher.function:123]: I think `DEPRECATED_REMOVED` is redundant with `TEST_DEPRECATED`, so the guard should be `#if !defined(MBEDTLS_USE_PSA_CRYPO) || !defined(MBEDTLS_TEST_DEPRECATED)`. That is, in plain English the condition for being able to use `cipher_setup_psa()` is "use-PSA is enabled, and we're allowed to test deprecated functions".

File: tests/suites/test_suite_cipher.function [hunk 4]
```diff
@@ -1072,7 +1074,8 @@ void test_vec_crypt( int cipher_id, int operation, data_t *key,
     memset( output, 0x00, sizeof( output ) );
 
     /* Prepare context */
-#if !defined(MBEDTLS_USE_PSA_CRYPTO)
+#if !defined(MBEDTLS_USE_PSA_CRYPTO) || defined(MBEDTLS_DEPRECATED_REMOVED) || \
```
Comment (mpg) [inline tests/suites/test_suite_cipher.function:1077]: Same as above.

**Review by mpg — CHANGES_REQUESTED** [review]
A few more dependency clean-ups and we'll be there.

File: tests/suites/test_suite_cipher.ccm.data [hunk 5]
```diff
@@ -479,385 +479,385 @@ depends_on:MBEDTLS_CAMELLIA_C:MBEDTLS_CCM_C
 auth_crypt_tv:MBEDTLS_CIPHER_CAMELLIA_128_CCM:"D75C2778078CA93D971F96FDE720F4CD":"003B8FD8D3A937B160B6A31C1C":"A4D499F78419728C19178B0C":"4B198156393B0F7796086AAFB454F8C3F034CCA966":"945F1FCEA7E11BEE6A2F":"":"9DC9EDAE2FF5DF8636E8C6DE0EED55F7867E33337D":0
 
 AES-128-CCM test vector NIST #1 PSA (P=0, N=7, A=0, T=4)
-depends_on:MBEDTLS_USE_PSA_CRYPTO:MBEDTLS_AES_C:MBEDTLS_CCM_C
+depends_on:MBEDTLS_USE_PSA_CRYPTO:!MBEDTLS_DEPRECATED_REMOVED:MBEDTLS_AES_C:MBEDTLS_CCM_C
```
Comment (mpg) [inline tests/suites/test_suite_cipher.ccm.data:482]: This should be changed to `MBEDTLS_TEST_DEPRECATED`, and in `gcm.data` as well.

**Review by mpg — CHANGES_REQUESTED** [review]
Looks almost good to me, just missed two data files I think.

**Review by mpg — APPROVED** [review]
LGTM - and the CI agrees :)

**Review by superna9999 — APPROVED** [review]

**Comment by mprse** [comment]
PR-Merge is pending, but looks green I think we can merge this one.
